src/testing/mocks/middleware/icache.ts

```diff
@@ -0,0 +1,73 @@
+/* tslint:disable:interface-name */
+import { create, invalidator } from '../../../core/vdom';
+import { MiddlewareResult } from '../../../core/interfaces';
+import Map from '../../../shim/Map';
+
+export function createICacheMock() {
+	const map = new Map<any, any>();
+	const factory = create({ invalidator });
+
+	const mockICacheFactory = factory(({ middleware: { invalidator } }) => {
```

Comment:
Do we need to recreate the `icache` middleware logic in the mock? It would be good if we could pass everything off to the real `icache` and only intercept the `set` call to get the promise (if it is one)?

There are examples in some of the other mock middleware that effectively do similar things, like the `resize` mock. Something like:

```ts
	const factory = create({ cache, invalidator });

	const mockICacheFactory = factory(({ id, middleware, properties, children }) => {
		const { callback } = icache();
		invalidate = middleware.invalidator;
		const icacheMiddleware = callback({
			id,
			middleware: { invalidator: middleware.invalidator, cache: middleware.cache },
			properties,
			children
		});
        return {
            getOrSet(key: any, value: any) {
                return icacheMiddleware.getOrSet(key, value);
            }
            get(key: any) {
               icacheMiddleware.get(key);
            }
            set(key: any, value: any) {
                // do things in here to get the promise to make it available to the mock
            }
            clear() {
               icacheMiddleware.clear();
            }
        }
	});
```

Reply:
Yep, seems possible.

Reply:
yeah that sounds good, I think we can even just monkey patch the icache that comes back so we'd only have to implement `set` and not deal with the rest? something like*:

```tsx
const factory = create({ cache, invalidator });
const mockICacheFactory = factory(({ id, middleware, properties, children }) => {
  const { callback } = icache();
  invalidate = middleware.invalidator;
  const icacheMiddleware = callback({ id, middleware: { invalidator: middleware.invalidator, cache: middleware.cache }, properties, children }); 
  const setter = icacheMiddleware.set;
  icaceMiddleware.set = (...args) => {
    // whatever
    return setter(...args);
  }
  return icacheMiddleware;
});
```
* t&c's apply, totally pseudo code 😄